### PR TITLE
Changes to README.md on main

### PR DESCRIPTION
**Pull request by LukeHollingsworth**
This PR includes changes I would like to be made to both the whittington_2020 branch and the main. 

**The only changes I would like to merge with the main branch are those to the README.md file.** These include options for develoment installations that work for both Unix and Windows OS; on my home PC (Windows), I was unable to use the existing command to install the development dependencies.

The rest of the changes should be made to whittington_2020 only. 

Thank you!

### CODE REVIEW

**Review by niksirbi — COMMENTED** [review]
Hey @LukeHollingsworth, thanks for updating the dev installation instructions. I think the split between `pip install -e '.[dev]'` and `pip install -e .[dev]` is not due to Windows vs Unix. 
It comes down to what shell you are using. 

On most shells (including `bash`) the `pip install -e .[dev]` syntax works fine. The exception is `zsh`, the default Mac shell. There,  square brackets `[` are interpreted as a globbing pattern, unless enclosed in quotes (see details [here](https://zsh.sourceforge.io/Doc/Release/Expansion.html#Filename-Generation)). 

Given that, you might say let's recommend `pip install -e '.[dev]'` for everyone, since it works on both `bash` and `zsh`. However, as you discovered, that sadly doesn't work on Windows (we can't have nice things).

Perhaps we can have `pip install -e .[dev]` as the default option, with a note that you should use `pip install -e `.[dev]` if you are using `zsh`.

Maybe phrased as follows:

If you want to contribute to the project, get the latest development version
from GitHub, and install it in editable mode, including the "dev" dependencies:

```bash
git clone https://github.com/SainsburyWellcomeCentre/NeuralPlayground/ --single-branch
cd NeuralPlayground
pip install -e .[dev]
```

> [!NOTE]
> If you are using the `zsh` shell (the default on MacOS), replace the last command with:
> ```zsh
> pip install -e `.[dev]`
> ```

**Comment by niksirbi** [comment]
Also I am a bit confused about this:
> This PR includes changes I would like to be made to both the whittington_2020 branch and the main.

If there are changes you want to make to `main`, but not to `whittington_2020`, the usual practice would be to open a separate branch (perhaps named sth like "update_readme") and open a PR from there.

**Comment by LukeHollingsworth** [comment]
Your solution to the shell-dependencies conflict sounds great! I just wanted to be sure that no one encountered a problem when first setting NeuralPlayground up on their local machine, which this should fix.

As for your confusion, it is most likely due my ignorance of pull requests. I think I should have generated 2 separate PRs, one for the changes to the whittington_2020 branch and then one for the main. Unfortunately, I didn't do this and so I was asking whether changes to the README.md file alone could be pulled into main, whilst everything else went into whittington_2020.

**Comment by niksirbi** [comment]
No worries at all, everyone has to be "new" at pull requests at some point.
Here is the workflow we use:

To simply add stuff to a branch that is not main, say `whittington_2020`, you don't need pull requests, you can simply push to that branch as you are used to.

The only protected branch is `main`. To make any change there:
- create a new "feature" branch off main
- commit and push stuff to that "feature" branch
- open a "Draft pull request" from the "feature" branch to main. Make sure to give a meaningful name to that pull request
- wait to see if the automated checks pass, fix any problems that arise
- once all checks are green and happy, mark that pull request as "Ready for review" (this will turn it from a "draft" pull request to a normal one). Request a review from someone on the team.
- Once the reviewer approves the pull request, either you or the reviewer may merge it into main. There are several options for this. We in the neuroinformatics unit prefer to use the "Squash and Merge" option. This means that all commits of the "feature" branch get squashed into a single commit message on the main branch. The title of the pull request becomes that commit message on main (hence the importance of a good name). This way the history on main remains clean and it is easy to find a past checkpoint should you need to go back to that

In general, the above workflow works best for small pull requests, with each one doing ideally a single thing. It's very tempting to mix many different things into one PR, but your future self will appreciate it if you keep things separated.

Some more details:
Often, while you are working on a "feature" branch, the main has progressed by some commits. This might create conflicts during the pull request. Best practice would be to merge the latest changes from main to the "feature" branch (can be done via `git merge main` while you are on "feature"), and resolve any arising conflicts. Another option would be rebasing, which is a bit trickier to understand, but luckily there are [good explainers online](https://www.thinktecture.com/en/tools/demystifying-git-rebase/).

I should put this and more in a comprehensive developer guide.

**Comment by niksirbi** [comment]
Regarding the current PR, I am happy to modify it so that only the README is changed. Alternatively,  you may want to close it and open a new one instead, if you prefer to practice the aforementioned workflow.

**Comment by niksirbi** [comment]
@LukeHollingsworth I think I've circumvented the CI issue of missing `torch` for now, by moving `torch` to the `dev` dependencies.

However, we are now up against the other issue I've highlighted above.

File: examples/agent_examples/tem_plot_test.py
Comment (ClementineDomine) [inline examples/agent_examples/tem_plot_test.py:1]: This probably should be deleted

File: examples/agent_examples/tem_run_test.py
Comment (ClementineDomine) [inline examples/agent_examples/tem_run_test.py:1]: Same here should this be kept ?

File: neuralplayground/agents/whittington_2020.py [hunk 1]
```diff
@@ -464,66 +484,79 @@ def collect_final_trajectory(self):
...
+            forward, tem, environments, include_stay_still=include_stay_still
+        )
+        analyse.zero_shot(forward, tem, environments, include_stay_still=include_stay_still)
+        analyse.location_occupation(forward, tem, environments)
+        self.g_rates, self.p_rates = analyse.rate_map(forward, tem, environments)
+        from_acc, to_acc = analyse.location_accuracy(forward, tem, environments)
+        return
+
+    def plot_rate_map(
+        self, rate_map_type=None, frequencies=["Theta", "Delta", "Beta", "Gamma", "High Gamma"], max_cells=30, num_cols=6
```
Comment (ClementineDomine) [inline neuralplayground/agents/whittington_2020.py:508]: Here I need to test it for different numbers and settings

File: neuralplayground/agents/whittington_2020.py [hunk 2]
```diff
@@ -464,66 +484,79 @@ def collect_final_trajectory(self):
...
+
         """
-        frequencies = ["Theta", "Delta", "Beta", "Gamma", "High Gamma"]
+
         figs = []
         axes = []
 
-        for i in range(5):
+        if rate_map_type == "g":
+            rate_maps = self.g_rates
```
Comment (ClementineDomine) [inline neuralplayground/agents/whittington_2020.py:525]: Can you explain this to me later ?

File: neuralplayground/arenas/batch_environment.py [hunk 3]
```diff
@@ -22,19 +22,21 @@ def __init__(self, environment_name: str = "BatchEnv", env_class: object = Simpl
                 Keyword arguments for the environment
```
Comment (ClementineDomine) [inline neuralplayground/arenas/batch_environment.py:22]: It seems that batch inherits from simple2D in thee nd

File: neuralplayground/arenas/discritized_objects.py [hunk 4]
```diff
@@ -339,32 +342,79 @@ def validate_action(self, pre_state, action, new_state):
             crossed_wall = crossed or crossed_wall
         return new_state, crossed_wall
 
-    # to be written again here
-    def plot_objects(self, history_data=None, ax=None, return_figure=False):
+    def plot_trajectory(
+        self,
```
Comment (ClementineDomine) [inline neuralplayground/arenas/discritized_objects.py:346]: Witch of the two plotting trajectory function do you use ?

File: tests/arena_exp_test.py [hunk 5]
```diff
@@ -156,3 +160,211 @@ def init_env(self):
...
+            batch_size=batch_size,
+            arena_x_limits=arena_x_limits,
+            arena_y_limits=arena_y_limits,
+            arg_env_params=discrete_env_params,
+        )
+        return [
+            env,
+        ]
+
+    def test_agent_interaction(self):
```
Comment (ClementineDomine) [inline tests/arena_exp_test.py:227]: Do you need he interaction test in both ?

Comment (LukeHollingsworth) [reply tests/arena_exp_test.py:227]: I think it's best yes. As we discussed, batch_environment should be a bit different from those that inherit from Simple2D and so I have put an interaction test in there with TEM as the agent.

File: neuralplayground/arenas/batch_environment.py (hunk 3 above)
Comment (LukeHollingsworth) [reply neuralplayground/arenas/batch_environment.py:22]: Where are you seeing this? I thought it is initialised using `Environment `in the batch_environment.py file and as `object` in the tests.

Comment (ClementineDomine) [reply neuralplayground/arenas/batch_environment.py:22]: ok it seems that you don't make use of it actually. I ll juste delete the env_class variable

**Review by ClementineDomine — APPROVED** [review]

**Pull request merged**